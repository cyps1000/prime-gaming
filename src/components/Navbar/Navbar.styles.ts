/**
 *  Material UI Imports
 * @see https://material-ui.com/styles/basics/
 */
import { makeStyles, Theme } from "@material-ui/core/styles";

/**
 * Styles the component
 */
const useStyles = makeStyles((theme: Theme) => ({
  root: {
    flexGrow: 1,
  },
  nav: {
    background: theme.palette.primary.main,
    boxShadow: "none",
  },
  actions: {
    display: "flex",
    alignItems: "center",
    flex: 1,
    justifyContent: "flex-start",
    marginRight: "2rem",
  },
  list: {
    width: 250,
  },
  paper: {
    background: theme.palette.primary.main,
  },
  menuIcon: {
    width: "1em",
  },
  copyright: {
    color: theme.palette.secondary.main,
    fontSize: 15,
    fontFamily: theme.typography.fontFamily,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    paddingTop: 20,
  },
  modal: {
    backgroundColor: theme.palette.primary.main,
  },
}));

export { useStyles };
