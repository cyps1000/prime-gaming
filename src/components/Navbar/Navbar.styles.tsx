/**
 *  Material UI Imports
 * @see https://material-ui.com/styles/basics/
 */
import { makeStyles, Theme, fade } from "@material-ui/core/styles";

/**
 * Styles the component
 */
const useStyles = makeStyles((theme: Theme) => ({
  root: {
    flexGrow: 1,
  },
  nav: {
    background: theme.palette.primary.main,
  },
  flag: {
    width: 20,
  },
  push: {
    marginLeft: "auto",
  },
  themeFix: {
    display: "flex",
    alignItems: "center",
    paddingLeft: 20,
    paddingRight: 30,
  },
  switchFix: {
    paddingLeft: 10,
  },
  logo: {
    maxWidth: 150,
  },
  search: {
    position: "relative",
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    "&:hover": {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    marginLeft: 0,
    marginRight: 15,
    width: "100%",
    [theme.breakpoints.up("sm")]: {
      marginLeft: theme.spacing(1),
      width: "auto",
    },
  },
  searchIcon: {
    padding: theme.spacing(0, 2),
    height: "100%",
    position: "absolute",
    pointerEvents: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  inputRoot: {
    color: "#FFFFFF",
  },
  inputInput: {
    padding: theme.spacing(1, 1, 1, 0),
    // vertical padding + font size from searchIcon
    paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
    transition: theme.transitions.create("width"),
    width: "100%",
    [theme.breakpoints.up("sm")]: {
      width: "12ch",
      "&:focus": {
        width: "20ch",
      },
    },
  },
}));

export { useStyles };
