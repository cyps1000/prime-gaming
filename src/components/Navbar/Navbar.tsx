import React, { useState } from "react";

/**
 * Imports Material UI Components
 */
import { useTheme } from "@material-ui/core/styles";
import useMediaQuery from "@material-ui/core/useMediaQuery";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import Drawer from "@material-ui/core/Drawer";
import List from "@material-ui/core/List";
import Divider from "@material-ui/core/Divider";
import ListItem from "@material-ui/core/ListItem";
import Typography from "@material-ui/core/Typography";

/**
 * Components Imports
 */
import Logo from "../Logo";
import NavbarSearchForm from "../NavbarSearchForm";
import ThemeSwitcher from "../ThemeSwitcher";
import LanguageSwitcher from "../LanguageSwitcher";
import NavbarUserMenu from "../NavbarUserMenu";
import NavbarMenu from "../NavbarMenu";
import NavbarUserMenuMobile from "../NavbarUserMenuMobile";
import NavbarMobileFooter from "../NavbarMobileFooter";
import FooterSocials from "../FooterSocials";
import RegisterModal from "../RegisterModal";
import LoginModal from "../LoginModal";
import ModalTitle from "../ModalTitle";

/**
 * Imports the component styles
 */
import { useStyles } from "./Navbar.styles";

/**
 * Defines the Modals' State interface
 */
interface ModalState {
  signUpModal: boolean;
  signInModal: boolean;
}

/**
 * Defines the props interface
 */
export interface NavbarProps {
  copyrightText?: string;
}

/**
 * Defines the default props
 */
const defaultProps: NavbarProps = {
  copyrightText: "Â© 2021 Prime Gaming",
};

/**
 * Displays the component
 */
const Navbar: React.FC<NavbarProps> = (props) => {
  const { copyrightText } = props;

  /**
   * Initializes the useTheme hook
   */
  const theme = useTheme();

  /**
   * Handles the tablet view
   */
  const isTablet = useMediaQuery(theme.breakpoints.down("md"));

  /**
   * Handles the mobile view
   */
  const isMobile = useMediaQuery(theme.breakpoints.down("xs"));

  /**
   * Gets the component styles
   */
  const classes = useStyles();

  /**
   * Initializes the Drawer state
   */
  const [open, setOpen] = useState(false);

  /**
   * Initializes the Modal state
   */
  const [modals, setModals] = useState({
    signUpModal: false,
    signInModal: false,
  } as ModalState);

  /**
   * Handles opening the modal
   */
  const openModal = (modalType: string) => {
    setModals((prevState) => {
      return { ...prevState, [modalType]: true };
    });
  };

  /**
   * Handles opening the Sign In Modal
   */
  const openSignInModal = () => openModal("signInModal");

  /**
   * Handles opening the Sign Up Modal
   */
  const openSignUpModal = () => openModal("signUpModal");

  /**
   * Handles closing the Sign In and Sign Up Modal
   */
  const closeModal = () =>
    setModals({ signUpModal: false, signInModal: false });

  /**
   * Handles opening the Drawer
   */
  const openDrawer = () => setOpen(true);

  /**
   * Handles closing the Drawer
   */
  const closeDrawer = () => setOpen(false);

  /**
   * Handles the mobile/tablet view
   */
  if (isMobile || isTablet) {
    return (
      <div className={classes.root} data-testid="navbar-mobile">
        <AppBar position="fixed" className={classes.nav}>
          <Toolbar>
            <NavbarMenu onClick={openDrawer} />
            <Logo />
          </Toolbar>
        </AppBar>
        <Drawer
          data-testid="navbar-drawer-mobile"
          open={open}
          onClose={closeDrawer}
          classes={{
            paper: classes.paper,
          }}
        >
          <div className={classes.list} data-testid="navbar-list-container">
            <List>
              <ListItem>
                <NavbarSearchForm withExpandAnimation={false} />
              </ListItem>
            </List>
            <Divider />
            <NavbarUserMenuMobile
              close={closeDrawer}
              openSignUpModal={openSignUpModal}
              openSignInModal={openSignInModal}
            />
            <Divider />
            <List>
              <ListItem button>
                <div className={classes.actions}>
                  <ThemeSwitcher />
                </div>
              </ListItem>
              <ListItem button>
                <div className={classes.actions}>
                  <LanguageSwitcher />
                </div>
              </ListItem>
            </List>
            <Divider />
            <NavbarMobileFooter close={closeDrawer} />
            <Divider />
            <FooterSocials className={classes.menuIcon} />
            <Divider />
            <Typography className={classes.copyright}>
              {copyrightText}
            </Typography>
          </div>
        </Drawer>
        <RegisterModal onClose={closeModal} open={modals.signUpModal}>
          <ModalTitle
            onClick={closeModal}
            classes={{
              container: classes.titleContainer,
              icon: classes.modalIcon,
            }}
          />
        </RegisterModal>
        <LoginModal onClose={closeModal} open={modals.signInModal}>
          <ModalTitle
            onClick={closeModal}
            classes={{
              container: classes.titleContainer,
              icon: classes.modalIcon,
            }}
          />
        </LoginModal>
      </div>
    );
  }

  return (
    <div className={classes.root} data-testid="navbar">
      <AppBar position="fixed" className={classes.nav}>
        <Toolbar>
          <Logo />
          <NavbarSearchForm withExpandAnimation />
          <div className={classes.actions}>
            <ThemeSwitcher />
            <LanguageSwitcher />
          </div>
          <NavbarUserMenu
            openSignUpModal={openSignUpModal}
            openSignInModal={openSignInModal}
          />
        </Toolbar>
      </AppBar>
      <RegisterModal onClose={closeModal} open={modals.signUpModal} />
      <LoginModal onClose={closeModal} open={modals.signInModal} />
    </div>
  );
};

Navbar.defaultProps = defaultProps;
export default Navbar;
