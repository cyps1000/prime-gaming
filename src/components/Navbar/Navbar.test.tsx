/**
 * @see https://testing-library.com/docs/react-testing-library/intro
 * @see https://www.robinwieruch.de/react-testing-library
 * @see https://www.smashingmagazine.com/2020/07/react-apps-testing-library/
 */

/**
 * Imports test utils
 */
import {
  render,
  changeViewport,
  userEvent,
  waitFor,
} from "../../utils/test-utils";

/**
 * External Imports
 */
import pretty from "pretty";

/**
 * Imports component
 */
import Navbar from "./Navbar";

/**
 * Navbar rendering tests
 */
describe("Navbar Rendering Tests", () => {
  it("renders the component without errors", () => {
    render(<Navbar />);
  });

  it("renders the default navbar when the screen size is above tablet", () => {
    const { getByTestId, rerender, queryByTestId } = render(<Navbar />);

    expect(getByTestId("navbar")).toBeInTheDocument();
    expect(queryByTestId("navbar-mobile")).not.toBeInTheDocument();

    window.matchMedia = changeViewport("lg");

    rerender(<Navbar />);

    expect(getByTestId("navbar")).toBeInTheDocument();
    expect(queryByTestId("navbar-mobile")).not.toBeInTheDocument();
  });

  it("renders the mobile navbar when the screen size is tablet or bellow", () => {
    const { getByTestId, rerender, queryByTestId } = render(<Navbar />);

    expect(getByTestId("navbar")).toBeInTheDocument();
    expect(queryByTestId("navbar-mobile")).not.toBeInTheDocument();

    window.matchMedia = changeViewport("md");

    rerender(<Navbar />);

    expect(getByTestId("navbar-mobile")).toBeInTheDocument();
    expect(queryByTestId("navbar")).not.toBeInTheDocument();

    window.matchMedia = changeViewport("sm");

    rerender(<Navbar />);

    expect(getByTestId("navbar-mobile")).toBeInTheDocument();
    expect(queryByTestId("navbar")).not.toBeInTheDocument();

    window.matchMedia = changeViewport("xs");

    rerender(<Navbar />);

    expect(getByTestId("navbar-mobile")).toBeInTheDocument();
    expect(queryByTestId("navbar")).not.toBeInTheDocument();
  });
});

/**
 * Navbar logic tests
 */
describe("Navbar Logic Tests", () => {
  it("while on mobile/tablet view, triggers a drawer when clicking on the navbar menu component", async () => {
    window.matchMedia = changeViewport("md");
    const { getByTestId } = render(<Navbar />);

    userEvent.click(getByTestId("navbar-menu-button"));

    expect(getByTestId("navbar-list-container")).toBeInTheDocument();
  });
});

/**
 * Navbar snapshot test
 */
describe("Navbar Snapshot Test", () => {
  it("passes the snapshot test", () => {
    const { container } = render(<Navbar />);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
