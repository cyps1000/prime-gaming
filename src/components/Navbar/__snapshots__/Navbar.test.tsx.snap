// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Snapshot Test passes the snapshot test 1`] = `
"<div class=\\"makeStyles-root-205\\" data-testid=\\"navbar\\">
  <header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-nav-206 mui-fixed MuiPaper-elevation4\\">
    <div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><img class=\\"makeStyles-logo-212\\" src=\\"logo.png\\" alt=\\"logo\\">
      <div class=\\"makeStyles-search-213\\">
        <div class=\\"makeStyles-searchIcon-214\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
            <path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path>
          </svg></div>
        <div class=\\"MuiInputBase-root makeStyles-inputRoot-215\\"><input placeholder=\\"Search...\\" type=\\"text\\" aria-label=\\"search\\" class=\\"MuiInputBase-input makeStyles-inputInput-216 makeStyles-expandAnimation-217\\" value=\\"\\"></div>
      </div>
      <div class=\\"makeStyles-actions-207\\">
        <div class=\\"makeStyles-container-218\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
            <path d=\\"M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z\\"></path>
          </svg><span class=\\"MuiSwitch-root makeStyles-switch-219\\"><span class=\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-222 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-223 Mui-checked\\" aria-disabled=\\"false\\"><span class=\\"MuiIconButton-label\\"><input class=\\"PrivateSwitchBase-input-225 MuiSwitch-input\\" name=\\"checkedB\\" type=\\"checkbox\\" aria-label=\\"primary checkbox\\" value=\\"dark-theme\\" checked=\\"\\"><span class=\\"MuiSwitch-thumb\\"></span></span><span class=\\"MuiTouchRipple-root\\"></span></span><span class=\\"MuiSwitch-track\\"></span></span><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
            <path d=\\"M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6c3.31 0 6 2.69 6 6s-2.69 6-6 6z\\"></path>
          </svg></div>
        <div data-testid=\\"language-switcher\\" class=\\"makeStyles-container-227\\"><img class=\\"makeStyles-flag-226\\" src=\\"https://www.countryflags.com/wp-content/uploads/romania-flag-png-xl.png\\" alt=\\"ro\\">
          <div><span class=\\"MuiSwitch-root\\"><span class=\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-222 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-223 Mui-checked\\" aria-disabled=\\"false\\" data-testid=\\"language-switch-input\\"><span class=\\"MuiIconButton-label\\"><input class=\\"PrivateSwitchBase-input-225 MuiSwitch-input\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"><span class=\\"MuiSwitch-thumb\\"></span></span><span class=\\"MuiTouchRipple-root\\"></span></span><span class=\\"MuiSwitch-track makeStyles-switchTrack-230\\"></span></span></div><img class=\\"makeStyles-flag-226\\" src=\\"https://www.countryflags.com/wp-content/uploads/united-kingdom-flag-png-large.png\\" alt=\\"uk\\">
        </div>
      </div>
      <div class=\\"makeStyles-menuItems-231\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Home</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">News</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">About</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Contact</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Sign In</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Sign Up</span><span class=\\"MuiTouchRipple-root\\"></span></button></div>
    </div>
  </header>
</div>"
`;
